heron/common/src/java/com/twitter/heron/common/basics/WakeableLooper.java

```diff
@@ -92,8 +92,13 @@ private void runOnce() {
   }
 
   private void onExit() {
-    for (Runnable r : exitTasks) {
-      r.run();
+    int s = exitTasks.size();
```

Comment:
I would suggest checking == 0, like

for (int i = 0; i < s; i ++) {
      exitTasks.get(i).run();
      // in case previous task cleared the list
      if (exitTasks.size() == 0) {
        break;
      }
}		    

Another qq, no tasks would be added when this function is called? I think onExit() might be safe but executeTasksOnWakeup() is a concern.

Reply:
Though is checking tasksOnWakeup.size() <= i more defensive programming?  With this check an index out of bounds exception cannot happen while for some reason or another not all the tasks are cleared you can get into a situation where i > list size causing index out of bounds exception

Reply:
it is more defensive, but confusing mathematically.

In fact I dont like either of them because it is not ideal to check the size an array when it can change.

Another option is to change the interface:
task.run() should return a boolean value indicate the loop should be continued or not. When it returns false, the loop should clear the list and return. Something like:

for (int i = 0; i < s; i ++) {
  if (!exitTasks.get(i).run()) {
    exitTasks.clear();
    break;
  }
}

This way we can avoid the clear() call in tasks.
How do you think?

Reply:
It seems that it is possible and allowed to add new tasks when the array is being executed. :(

but i think the new tasks added by the tasks before restore should be cleared too. So the clear() call is still correct.